Adds the ability to pin the currently developed photo to the 2nd darkroom window

masterpiga opened the pull request:
This change adds two buttons to the 2nd darkroom window, making this functionality more generally useful:

1. A "pin" button to keep the 2nd window focused on the currently developed images. Changing images in the darktable main window does not affect the content of the 2nd window. This is useful to use the 2nd window as a reference, e.g., for style matching or color grading.

2. A "keep on top" button, which keeps the 2nd window on top of the main one. This is especially useful for macOS users, as the OS does not implement this functionality natively.

The two changes are bundled together because on macOS pinning the image without being able to keep it in on top greatly reduces its usefulness. Additionally, the code footprint of (2) is negligible.

The change was proposed and discussed on [Pixls](https://discuss.pixls.us/t/new-feature-display-different-image-in-second-darkroom-window/54633), and the community seemed to welcome it.

The implementation discussed on Pixls was taking a full-size snapshot of the image. Following @dterrahe's suggestions, I reimplemented it to use a separate pixelpipe for the 2nd window.

When the image is pinned, the current pipe is duplicated and the view of 2nd window updated to point to insist on the new pipe. When the image is unpinned, the 2nd pipe is destroyed and the main darkroom pipe is restored.

See demo video below:

https://github.com/user-attachments/assets/01218baa-d8b8-4c06-8837-6be2cd846a7d

You will see a lot of examples with composite (overlay) and watermark, as the changes introduced in develop.c to be able to reuse the pipe had broken that functionality. It took me quite a bit to track down the issue and fix it. The problematic bit is documented in the code.

I tested extensively opening and closing the 2nd window, pinning and unpinning repeatedly while the main image is still being rendered, which led me to fix some issues with the 2nd window cleanup code.

I am quite confident that the logic is sound now, but I am not familiar with the pixelpipe code so a close inspection from more expert eyes is very welcome. Everything seems to work as expected, but you never know.

masterpiga committed:
Adds two toggle buttons to the 2nd window
- Pin/unpin the image currently displayed in the main view to the 2nd window. It is possible to pin at image at any point of the history.
- Keep the 2nd window on top (especially useful for macOS, whose WM does not have this functionality).

It uses a separate pixelpipe for the 2nd window.
When the image is pinned, the current pipe is duplicated and the view of 2nd window updated to point to insist on the new pipe.
When the image is unpinned, the 2nd pipe is destroyed and the main darkroom pipe is restored.

TurboGit reviewed (changes requested):
Tested today. I have two requests:

1. The "stays on top" does not work on GNU/Linux
2. I'd like the two buttons to hide themself after some time and be drawn again when the mouse is close. Something similar is done on the histogram.

masterpiga committed:
Make 2nd window buttons appear when the 2nd window is hovered.

masterpiga committed:
- Make 2nd window a transient of the main window,
  so that it always stays on top, and drop the
  dedicated "keep on top" button.

- Implement focus following so that tooltips on
  the 2nd window show also if the window is not
  explicitly focused.

masterpiga commented:
Done, @TurboGit.

This is on Wayland:

https://github.com/user-attachments/assets/9dfeaafe-36a6-46ac-8c46-8a1b05352411

Tested also on macOS and X11.

- 2nd window stays always on top via `transient_for`
- Focus follows the mouse to enable tooltip display with no need to click the 2nd window
- Pin button only shows when 2nd window focused/hovered

TurboGit reviewed (approved):
Works for me, thanks!

anoderay commented:
When trying this out for documenting the feature I noticed that with slower processing the second window behaves somewhat erratically (OpenCL wasn't switched to fast mode). While in the main window behaves sluggish:
1.) When switching to a new photo (especially with heavy processing) the second window will stick to the previous photo until the main preview has finished rendering and only then update. 
2.) Photo in portrait mode will first render in horizontal orientation and then rotate correctly (this happens with faster processing, too but much quicker).

Not sure if this is fixable because the second windows preview uses it's own pipeline but wanted to track this here anyway. 

Also when closing and opening dt the second window needs to be opened every time - not sure whether that is intended? I suppose for multi-monitor users it might be helpful to have an option to keep the second window open across program launches?

masterpiga commented:
Thanks for the feedback, @anoderay!

> Also when closing and opening dt the second window needs to be opened every time - not sure whether that is intended? I suppose for multi-monitor users it might be helpful to have an option to keep the second window open across program launches?

Yes, you are right, there is margin for streamlining the UX there. The initial change was minimalistic on purpose, I am planning to iterate on it to smooth these friction points, also based on user feedback when the feature goes mainstream after the next release.
